pkg/remote/codec/default_codec.go

```diff
@@ -150,6 +190,54 @@ func (c *defaultCodec) EncodeMetaAndPayload(ctx context.Context, message remote.
 	return nil
 }
 
+// encodeMetaAndPayloadWithCRC32C encodes payload and meta with crc32c checksum of the payload.
+func (c *defaultCodec) encodeMetaAndPayloadWithCRC32C(ctx context.Context, message remote.Message, out remote.ByteBuffer, me remote.MetaEncoder) error {
+	var err error
+
+	// 1. encode payload and calculate crc32c checksum
+	payloadOut := netpoll.NewWriterByteBuffer(netpoll2.NewLinkBuffer())
+	if err = me.EncodePayload(ctx, message, payloadOut); err != nil {
+		return err
+	}
+	// get the payload from buffer
+	payload, payloadLen, err := payloadOut.(remote.NocopyRead).GetBytesNoCopy()
+	if err != nil {
+		// release if err
+		payloadOut.Release(err)
+		return err
+	}
+	crc32c := getCRC32C(payload)
+	strInfo := message.TransInfo().TransStrInfo()
+	if crc32c != "" && strInfo != nil {
+		strInfo[transmeta.HeaderCRC32C] = crc32c
+	}
+	// set payload length before encode TTHeader.
+	message.SetPayloadLen(payloadLen)
+
+	// 2. encode header and return totalLenField if needed
+	// In this case, set total length during TTHeader encode
+	if _, err = ttHeaderCodec.encode(ctx, message, out); err != nil {
+		return err
+	}
+
+	// 3. write payload to the buffer after TTHeader
+	if netpoll.IsNetpollByteBuffer(out) {
+		// append buffer only if the input buffer is a netpollByteBuffer
+		// release will be executed in AppendBuffer
+		err = out.AppendBuffer(payloadOut)
+	} else {
+		// convert [][]byte to []byte
+		p := flatten2DSlice(payload, payloadLen)
```

Comment:
copy once here and another time in `out.WriteXXX` ?

Reply:
Maybe it's better to use a for loop here to call `out.WriteXXX` on each segment?

Reply:
call writeDirect only one time will be better. Otherwise,  using loop will increase the time of object allocation in writeXXX.